### PR TITLE
[DNM] Artifact Weapons

### PR DESCRIPTION
**Pull request by LordFowl**
Secret chems can now metabolize with a philosopher's stone and rare minerals to create an enchantment rune. These enchantment runes can be applied to a specific item to create an artifact. Only one of any type of artifact can be created at any time.

Alternatively, these artifacts can be rarely found in xenoarchaeology. Should the xenoarch artifacts be discovered the chemical recipes will no longer work for the respective artifact.

Also adds zweihanders, for da bass cannon:

**Zweihanders:** Special material weapons. They can be effectively wielded with one hand or two hands. A one-handed zweihander does the same damage as a two handed zweihander, but has the cleaving effect. A two handed zweihander possesses the reach effect and an improved armor penetration.

**Artifacts:**
Aegis: The Aegis is the perfect shield of Perseus' age. It has perfect projectile projection, and somewhat above standard melee projection. Each attack it blocks increments an absorption counter. Activating the shield heals the user's physical wounds for the value of this absorption counter.

Horn of Plenty: The Horn of Plenty is the Cornucopia of ancient myth. Any reagent inside of the Horn of Plenty will never run out. The Horn of Plenty can be emptied by activating it in hand, as if you were sealing a beaker.

Götterdämmerung: The ancient sword of the jotunn Surtr, this flamberge is the ultimate form of a zweihander, with improved damage and armor penetration. Furthermore, the sword is always on fire.

Cloak of Invisibility: Simply confers invisibility upon the user.
~~Deck of Many Things: The deck of many things is equipped with all the major and minor arcana of a tarot deck. Each major arcana has a special major effect of negative, neutral, or positive effect. Each suit of minor arcana has a variable value of typical positive value.~~

Sunlight Scroll: The sunlight scroll tells the legend of a man pure of heart and jolly in his co-operation. Recalling this legend summons forth a beam of tangible sunlight, infused with energy and heat. This beam can be thrown as a javelin for great effect, particularly upon the supernatural.

**To-do:**

~~- [ ] Finish the Deck of Many Things~~
- [x] Finish sprites. Maybe consult an actual spriter.
- [x] Consider ways of making the Götterdämmerung more interesting than a really good energy sword.
- [x] Xenoarch integration
- [ ] Testing.
- [ ] Balance.

### CODE REVIEW

File: code/modules/spells/artifacts.dm [hunk 1]
```diff
@@ -150,18 +150,550 @@
...
+	factor_paths = list(/obj/item/weapon/reagent_containers, /obj/item/weapon/reagent_containers/food/drinks)
+
+/obj/item/weapon/reagent_containers/glass/cornucopia
+	name = "Horn of Plenty"
+	desc = "In some cultures it is known as the Cornucopia. Fill it with any reagent and it will never run dry. Whisper to it its power word to empty it."
+	icon = 'icons/obj/wizard.dmi'
+	icon_state = "cornucopia"
+
+/obj/item/weapon/reagent_containers/glass/cornucopia/attack_self()
+	usr << "<span class='notice'>You whisper to [src] its power word, and it drains itself dry.</span>"
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: Illegal use of `usr`. This proc has a `user` arg.

File: code/modules/spells/artifacts.dm [hunk 2]
```diff
@@ -150,18 +150,550 @@
...
+		return 1
+
+	if(target.reagents && !target.reagents.get_free_space())
+		user << "<span class='notice'>[target] is full.</span>"
+		return 1
+
+
+
+	var/contained = reagentlist()
+	target.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been splashed with [name] by [user.name] ([user.ckey]). Reagents: [contained]</font>")
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: Use the attack log procs.

File: code/modules/spells/artifacts.dm [hunk 3]
```diff
@@ -150,18 +150,550 @@
...
+		user << "<span class='notice'>\The [src] is empty.</span>"
+		return 1
+
+	//var/types = target.find_type()
+	var/mob/living/carbon/human/H
+	if(istype(target, /mob/living/carbon/human))
+		H = target
+
+	if(target == user)
+		if(istype(user, /mob/living/carbon/human))
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: `ishuman(user)`

File: code/modules/spells/artifacts.dm [hunk 4]
```diff
@@ -150,18 +150,550 @@
...
+	if(!istype(target))
+		return 0
+
+	if(!reagents || !reagents.total_volume)
+		user << "<span class='notice'>\The [src] is empty.</span>"
+		return 1
+
+	//var/types = target.find_type()
+	var/mob/living/carbon/human/H
+	if(istype(target, /mob/living/carbon/human))
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: `ishuman(target)`

File: code/modules/spells/artifacts.dm [hunk 5]
```diff
@@ -150,18 +150,550 @@
...
+				user << "<span class='warning'>\The [blocked] is in the way!</span>"
+				return
+
+		user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN) //puts a limit on how fast people can eat/drink things
+		self_feed_message(user)
+		target.reagents.add_reagent(reagents, amount_per_transfer_from_this)
+		feed_sound(user)
+		return 1
+	else
+		if(istype(target, /mob/living/carbon/human))
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: `ishuman(target)`

File: code/modules/spells/artifacts.dm [hunk 6]
```diff
@@ -150,18 +150,550 @@
...
+		other_feed_message_start(user, target)
+
+		user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+		if(!do_mob(user, target))
+			return
+
+		other_feed_message_finish(user, target)
+
+		var/contained = reagentlist()
+		target.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been fed [name] by [user.name] ([user.ckey]). Reagents: [contained]</font>")
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: Attack log procs.

File: code/modules/spells/artifacts.dm [hunk 7]
```diff
@@ -150,18 +150,550 @@
...
+	sharp = 1
+	attack_verb = list("attacked", "chopped", "cleaved", "torn", "cut")
+	default_material = "steel"
+	wielded_ap = 60
+	unwielded_ap = 20
+	light_color = "#ED9200"
+	light_power = 1
+	light_range = 4
+
+/obj/item/weapon/material/twohanded/zweihander/gotterdammerung/process()
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: Does this actually start processing anywhere?

File: code/modules/spells/artifacts.dm [hunk 8]
```diff
@@ -150,18 +150,550 @@
...
+	throwforce = 20
+	throw_speed = 10
+	throw_range = 30
+	damtype = "fire"
+	light_color = "#FDD023"
+	light_power = 1
+	light_range = 4
+
+/obj/item/weapon/sunlight_spear/Initialize()
+	..()
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: `. = ..()`

File: code/modules/spells/artifacts.dm [hunk 9]
```diff
@@ -150,18 +150,550 @@
...
+	QDEL_IN(src, 30 SECONDS) //meant to throw it, not stab with it
+
+/obj/item/weapon/sunlight_spear/Destroy()
+	playsound(user, 'sound/magic/lightningbolt.ogg', 20, 1)
+	..()
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: `return ..()`

File: code/modules/spells/artifacts.dm [hunk 10]
```diff
@@ -150,18 +150,550 @@
...
+		L.apply_effect(2,INCINERATE)
+		L.electrocute_act(10, tesla_shock = 1)
+		if(L.find_type() & TYPE_WEIRD)
+			L.apply_effect(2,INCINERATE)
+			L.adjustFireLoss(5)
+
+/obj/item/weapon/sunlight_spear/throw_at()
+	..()
+	var/obj/item/weapon/sunlight_scroll/S
+	if(user.l_hand && istype(user.l_hand,/obj/item/weapon/sunlight_scroll))
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: `user.l_hand` check is redundant, `istype(null)` is always false.

File: code/modules/spells/artifacts.dm [hunk 11]
```diff
@@ -150,18 +150,550 @@
...
+		S = user.l_hand
+	else if(user.r_hand && istype(user.r_hand,/obj/item/weapon/sunlight_scroll))
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: Same here.

File: code/modules/spells/artifacts.dm [hunk 12]
```diff
@@ -150,18 +150,550 @@
...
+			E.damage += rand(10, 11)
+			if(E.damage > 12)
+				M.eye_blurry += rand(3,6)
+			if (E.damage >= E.min_broken_damage)
+				M.sdisabilities |= BLIND
+			else if (E.damage >= E.min_bruised_damage)
+				M.eye_blind = 5
+				M.eye_blurry = 5
+				M.disabilities |= NEARSIGHTED
+				spawn(100)
```
Comment (Lohikar) [inline code/modules/spells/artifacts.dm]: Timer. There's already a proc defined on mob for clearing NEARSIGHTED you could use.

File: code/modules/spells/artifacts.dm [hunk 13]
```diff
@@ -150,18 +150,550 @@
...
+		var/mob/living/L = A
+		L.apply_effect(rand(4,12),INCINERATE)
+		L.adjustFireLoss(force/2)
+
+/obj/item/enchantment/singularity
+	name = "chaos enchantment rune"
+	desc = "A scroll marked with an ancient rune for chaos. On the reverse side is a glyph representing a playing card."
+	icon_state = "singularity"
+	product = /obj/item/weapon/singularity_deck
+	factor_paths = list(/obj/item/weapon/deck, /obj/item/weapon/deck)
```
Comment (TheGreatJorge) [inline code/modules/spells/artifacts.dm]: This list seems to have two same values in it.

Comment (LordFowl) [reply code/modules/spells/artifacts.dm]: Each one is distinct in its own special way.

**Comment by OneOneThreeEight** [comment]
i'm [redacted] the abysswalker and this is my favorite PR in Lordran.
Genuinely looks cool though. Alchemy soon.

**Pull request closed without merging**